Tutorial: The guide to frontend vs backend group analytics implementations (#4642)
PostHog/posthog.com: contents/tutorials/frontend-vs-backend-group-analytics.md

---

> title: The guide to frontend vs backend group analytics implementations

Reviewer: ```suggestion
title: Understanding group analytics: frontend vs backend implementations
```

Reviewer: Think this feels a bit more elegant

    '$groups' => array("company" => "posthog")
));
```

```segment
analytics.track('login', {
    "$groups": {
        "company": "posthog"
    }
})
```

</MultiLanguage>

You also can’t update the group properties from the capture call. You must use a separate `group_identify` which looks like this:

<MultiLanguage>

```python
posthog.group_identify('company', 'posthog', {
    'company_name': 'PostHog',
    'score': 99
})
```

```go
client.Enqueue(posthog.GroupIdentify{
    Type: "company",
    Key:  "posthog",
    Properties: posthog.NewProperties().
        Set("company_name", "PostHog").
        Set("score", 99),
})
```

```node
posthog.groupIdentify({
    groupType: 'company',
    groupKey: 'posthog',
    properties: {
        company_name: 'PostHog',
        score: 99
    }
})
```

```php
PostHog::groupIdentify(array(
    'groupType' => 'company',
    'groupKey' => 'posthog',
    'properties' => array("company_name" => "PostHog", "score" => 99)
));
```

```segment
analytics.track('$groupidentify', {
    "$group_type": "company",
    "$group_key": "posthog",
    "$group_set": {
        "company_name": "PostHog",
        "score": 99
    }
})
```

</MultiLanguage>

> PostHog currently stores groups separately from events, but soon, we will merge them. This means calling `group_identify` will not update the group details for past events (like it does now). To handle this, we recommend calling `group_identify` whenever group properties change. This makes sure events always have the most updated group properties.

Reviewer: ```suggestion
PostHog currently stores groups separately from events, but soon, we will merge them. This means calling `group_identify` will not update the group details for past events (like it does now). To handle this, we recommend calling `group_identify` whenever group properties change. This ensures events always have the most updated group properties.
```

The benefit of the backend is that you don’t have to call `reset()` when you’re done because you have no session to reset. If you’ve set up your capture calls properly, they should be sending data from the active user or group (which will be none).

## Next steps in group analytics

> This should give you a better understanding of how you need to implement frontend vs backend group analytics From here, you can utilize your group analytics to [view groups and their properties](/manual/group-analytics#viewing-groups-and-their-properties), [analyze group insights](/manual/group-analytics#analyzing-group-insights), and [setup group feature flags](/manual/group-analytics#integrating-groups-with-feature-flags). Check out [tracking how teams use your product](/tutorials/tracking-teams) for more ideas.

Reviewer: ```suggestion
This should give you a better understanding of how you need to implement frontend vs backend group analytics. From here, you can utilize your group analytics to [view groups and their properties](/manual/group-analytics#viewing-groups-and-their-properties), [analyze group insights](/manual/group-analytics#analyzing-group-insights), and [setup group feature flags](/manual/group-analytics#integrating-groups-with-feature-flags). 

Check out [tracking how teams use your product](/tutorials/tracking-teams) for more ideas.
```

---
title: The guide to frontend vs backend group analytics implementations
sidebar: Docs
showTitle: true
author: ['ian-vanagas']
date: 2022-11-09
featuredImage: ../images/tutorials/banners/frontend-vs-backend-group-analytics.png
topics: ["configuration"]
---

Group analytics is a powerful feature for understanding how groups such as organizations, customers, and companies use your product as a unit. It provides a new level of analysis between individual users and all your users.

When implementing group analytics, there are some differences to look out for between the frontend implementation and the backend one. This tutorial goes over why there are differences and how to implement group analytics on the frontend compared to the backend.

## Why are there differences

Before we go into the implementation, it is useful to understand **why** there are differences between the frontend and backend. 

> A group is generally defined as a collective of people with a shared identity. At PostHog, we think about (and store) everything as **events**. This means a group for us is a set of events related to a group identity (and properties). People (in PostHog) are the same, a set of events related to a person identity (and properties). 

Reviewer 2: > A group is generally defined as a collective of people with a shared identity.

Hmm, this isn't correct. A group can be both a collective of people, OR, multiple groups can belong to the same person, which is why groups are based on events.

To give an example: Organisation is a group, but so is a paymentID.

Every time a person makes a payment, a new 'group' paymentID is generated. You can see how paymentID is not a collective of people, but a collective of paymentIDs makes a person. Makes sense?

Author: I'm trying to say in this paragraph that how normal people think of a group is a collective of people, but at PostHog we think of a group as a collective of events.

I get that a person can belong to multiple groups, but I don't understand how that changes the "general" definition of a group. The people of an organization share an identity, so are a group. The events share a paymentID so they are a group too. 

Does changing the sentence to "A group is generally defined as a collective of people or things with a shared identity." make it more correct?

Reviewer 2: Ahhh, I see, what you're defining is a 'group' in english, not a posthog group. It adds more confusion honestly, since I'd _hope_ people reading this know what an english 'group' is already?

I'd make what you're intending to say explicit, or just get rid of the first line? Thoughts?

Reviewer 2: ```suggestion
At PostHog, we think about (and store) everything as **events**. This means a group for us is a set of events related to a group identity (and properties). People (in PostHog) are the same, a set of events related to a person identity (and properties). 
```

Reviewer 2: Also, what is the intention behind `(and properties)` ? 

"This means a group for us is a set of events with a common group identity. People in PostHog are the same, a set of events with a common person identity."

Author: Good points, I'll make both changes.

    'company_name': 'PostHog',
    'score': 99
})
```

```go
client.Enqueue(posthog.GroupIdentify{
    Type: "company",
    Key:  "posthog",
    Properties: posthog.NewProperties().
        Set("company_name", "PostHog").
        Set("score", 99),
})
```

```node
posthog.groupIdentify({
    groupType: 'company',
    groupKey: 'posthog',
    properties: {
        company_name: 'PostHog',
        score: 99
    }
})
```

```php
PostHog::groupIdentify(array(
    'groupType' => 'company',
    'groupKey' => 'posthog',
    'properties' => array("company_name" => "PostHog", "score" => 99)
));
```

```segment
analytics.track('$groupidentify', {
    "$group_type": "company",
    "$group_key": "posthog",
    "$group_set": {
        "company_name": "PostHog",
        "score": 99
    }
})
```

</MultiLanguage>

PostHog currently stores groups separately from events, but soon, we will merge them. This means calling `group_identify` will not update the group details for past events (like it does now). To handle this, we recommend calling `group_identify` whenever group properties change. This makes sure events always have the most updated group properties.

> The benefit of the backend is that you don’t have to call `reset()` when you’re done because you have no session to reset. If you’ve set up your capture calls properly, they should be sending data from the active user or group (which will be none).

Reviewer 2: Optional whether you want to include or not, just more context:

There are no sessions on the backend because usually backends deal with multiple requests from different people one after the other. Thus it becomes very awkward to enforce creating a session that lasts only for one request, and reduces flexibility for implementers.

Author: added, thanks!
